### DISCUSSION TITLE
Single Project

### DISCUSSION THREAD

**Discussion by marcosgerene** [AvaloniaUI/Avalonia#17844] [Q&A] [answered]
Hi folks,

In April UNO Project announced Single Project (https://platform.uno/blog/the-first-and-only-true-single-project-for-mobile-web-desktop-and-embedded-in-net/). MAUI does the same (although MAUI is a joke).

Is there something like that in Avalonia? Is it even in the roadmap?

---

Edit: I have found it: https://x.com/AvaloniaUI/status/1857310602923765833

![image](https://github.com/user-attachments/assets/f772508a-218f-4d4d-9758-37e59eb1c1c2)

**Comment by MikeCodesDotNET** [comment]
I've been using it with [XPF](https://avaloniaui.net/xpf) for ~6 months, so I'm 99.9% we can support it in vanilla Avalonia. We'd just need to update the project templates. I'll leave it with @maxkatz6 to decide if we want to support it officially.

**Reply by thevortexcloud** [reply]
From memory it _does_ work (I have not tried it myself). It's just not documented anywhere. I seem to recall it's even being used in a couple places for some tests. EG there is a sample project here https://github.com/AvaloniaUI/Avalonia/tree/c36493d1deccb0b9dc32fb8c631e94d3b9cfdeaf/samples/SingleProjectSandbox

**Reply by maxkatz6** [reply]
Yes, it is supported in a decent way but not documented. A proof of concept in a way.

By "decent way" I mean that users need to reference platform specific packages in their single project. I.e. `Avalonia.iOS` or `Avalonia.Android` depending on the current TargetFramework.
It can be avoided by either using a workload (what old MAUI did) or using a special SDK (what Uno does, or what newer MAUI does, but through standard .NET SDK). We are not interested in maintaining neither of these two approaches, as `single project` is not worth the effort.

XPF comes with a custom .NET SDK either way, so it was pretty easy to enable there.

**Reply by maxkatz6** [reply, marked as answer]
For implementation details you can see this PR https://github.com/AvaloniaUI/Avalonia/pull/9613

To enable this, you need to add `<AvaloniaSingleProject>true</AvaloniaSingleProject>` to your project, add all necessary target frameworks (that are supported by the .NET SDK at least) and then enable package references conditionally:
```xml
<PackageReference Include="Avalonia.iOS" Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'ios'" />
<PackageReference Include="Avalonia.Android" Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'" />
<PackageReference Include="Avalonia.Browser" Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'browser'" />
<!-- Plain "net9.0" can be used for desktop -->
<PackageReference Include="Avalonia.Desktop" Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == ''" />
```

Everything else would work as how typical single project support is in other .NET frameworks. You need to create platform specific folders (Platform/Android) and move bootstrap code (like, MainActivity.cs or Program.cs) in there. [Like it was done here](https://github.com/AvaloniaUI/Avalonia/tree/c36493d1deccb0b9dc32fb8c631e94d3b9cfdeaf/samples/SingleProjectSandbox/Platforms).

**Reply by MikeCodesDotNET** [reply]
> We are not interested in maintaining neither of these two approaches, as single project is not worth the effort.

I have to agree with @maxkatz6 on this. 

Single project for any reasonably complex application is a nightmare, compared to having a clear separation of concerns. For me, it always felt like an exercise in marketing, rather than a genuinely useful feature.

**Reply by marcosgerene** [reply]
>It can be avoided by either using a workload (what old MAUI did) or using a special SDK (what Uno does, or what newer MAUI does, but through standard .NET SDK). We are not interested in maintaining neither of these two approaches, as single project is not worth the effort.

@MikeCodesDotNET maybe it could be part of something as your paid tools. Single project is not vital but I fell more comfortable with it.


>Single project for any reasonably complex application is a nightmare, compared to having a clear separation of concerns. For me, it always felt like an exercise in marketing, rather than a genuinely useful feature.

I understand your point, but ask community on twitter and after that "ignore" the result does not sound well, maybe remove the pool or just post why you guys decided to not go in this direction could be better.


Anyway, thank you for your time @MikeCodesDotNET  and @maxkatz6
